frontend/views/utils/captureLogs.js

```diff
@@ -0,0 +1,102 @@
+import sbp from '~/shared/sbp.js'
+import { CAPTURED_LOGS } from '~/frontend/utils/events.js'
+
+const SIZE_LIMIT_KB = 100
+const isNewSession = !sessionStorage.getItem('NEW_SESSION')
+const initialLogs = localStorage.getItem(CAPTURED_LOGS)
+
+// Mark in localStorage that is a new visit (and session)
+if (isNewSession) { sessionStorage.setItem('NEW_SESSION', 1) }
+
+const setNewEntry = [
+  ...(JSON.parse(initialLogs) || []),
+  {
+    type: isNewSession ? 'NEW_SESSION' : 'NEW_VISIT',
+    msg: Date.now()
+  }
+]
+
+localStorage.setItem(CAPTURED_LOGS, JSON.stringify(setNewEntry))
+
+// Create a middleware to capture new console logs...
+const giConsole = (function (csl) {
+  return {
+    log: function (...args) {
+      csl.log(...args)
+    },
+    info: function (...args) {
+      csl.info(...args)
+    },
+    warn: function (...args) {
+      csl.warn(...args)
+      captureMsg('warn', ...args)
+    },
+    error: function (...args) {
+      csl.error(...args)
+      captureMsg('error', ...args)
+    },
+    debug: function (...args) {
+      csl.debug(...args)
+      captureMsg('debug', ...args)
+    }
+  }
+}(window.console))
+
+function captureMsg (type, ...msg) {
+  const logs = localStorage.getItem(CAPTURED_LOGS)
+  const logsParsed = JSON.parse(logs) || []
+
+  // Reference: https://stackoverflow.com/a/15720835/4737729
+  // REVIEW: Cut the log from a better place? Maybe since last session?
+  var size = logs ? logs.length * 2 / 1024 : 0
+  if (size > SIZE_LIMIT_KB) {
```

Comment:
@taoeffect What type of cleanup do you suggest to do?

LS doesn't provide any direct API to know its size or to delete old data. It's totally up to us.
I think looking for the latest `NEW_SESSION` entry and delete everything before that can be a good solution.

LS size limit is 9Mb, do you think that's too much? I didn't test this up to 1Mb and it was okay.

Reply:
@taoeffect the log itself is saved as a string - a single line. and when parsed it's an array of logs. So we can't "detect" the number of lines. But we can detect the **string length** or the **array length**.
So... what do you think it's the best string length limit? Based on that, we can remove 1/4 of the arrays (the oldest).

Reply:
Did you see the algorithm I outlined below that's able to detect number of lines because of how it stores them?